File: triangle.cpp
#include "triangle.hpp"

Triangle::Triangle(const Point &p0, const Point &p1, const Point &p2, const Vector &color)
    : p0(p0), p1(p1), p2(p2), color(color)
{
    calculateNormal();
}

bool Triangle::intersect(const Point &rayOrigin, const Vector &rayDirection, double &t) const
{
    // Verifica a interseção com o plano que contém o triângulo
    Plane containingPlane = getContainingPlane();
    if (!containingPlane.intersect(rayOrigin, rayDirection, t))
    {
        return false; // Não há interseção com o plano
    }

    // Calcula o ponto de interseção no plano
    Point intersectionPoint = rayOrigin.translate(rayDirection.scale(t));

    // Calcula as coordenadas baricêntricas
    Vector edge0 = p1.difference(p0);
    Vector edge1 = p2.difference(p0);
    Vector intersectionVec = intersectionPoint.difference(p0);

    double dot00 = edge0.dot(edge0);
    double dot01 = edge0.dot(edge1);
    double dot11 = edge1.dot(edge1);
    double dotP0 = intersectionVec.dot(edge0);
    double dotP1 = intersectionVec.dot(edge1);

    double invDenom = 1.0 / (dot00 * dot11 - dot01 * dot01);

    // Calcula os valores de alpha, beta e gamma
    double alpha = (dot11 * dotP0 - dot01 * dotP1) * invDenom;
    double beta = (dot00 * dotP1 - dot01 * dotP0) * invDenom;
    double gamma = 1.0 - alpha - beta;

    // Verifica se os valores estão dentro dos limites adequados
    return (alpha >= 0.0 && beta >= 0.0 && gamma >= 0.0 && alpha <= 1.0 && beta <= 1.0 && gamma <= 1.0);
}

void Triangle::calculateNormal()
{
    Vector edge1 = p1.difference(p0);
    Vector edge2 = p2.difference(p0);
    normal = edge1.cross(edge2).normalize();
}

Plane Triangle::getContainingPlane() const
{
    return Plane(p0, normal, color);
}

File: src/triangle.py
from src.plane import Plane
from src.point import Point
from src.shape import Shape
from src.vector import Vector


class Triangle(Shape):
    def __init__(self, p0: Point, p1: Point, p2: Point, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self.p0 = p0
        self.p1 = p1
        self.p2 = p2
        self.normal = self._calculate_normal()
        self.containing_plane = self.get_containing_plane()

    def _calculate_normal(self) -> Vector:
        v0 = self.p1.difference(self.p0)
        v1 = self.p2.difference(self.p0)
        return v0.cross(v1).normalize()

    def get_containing_plane(self) -> Plane:
        return Plane(self.p0, self.normal, self.color)

    def intersect(self, ray_origin: Point, ray_direction: Vector) -> tuple[bool, float]:
        intersect, t = self.containing_plane.intersect(ray_origin, ray_direction)
        if not intersect:
            return False, 0

        intersection_point = ray_origin.translate(ray_direction.scale(t))

        v0 = self.p1.difference(self.p0)
        v1 = self.p2.difference(self.p0)
        intersection_vector = intersection_point.difference(self.p0)

        dot00 = v0.dot(v0)
        dot01 = v0.dot(v1)
        dot11 = v1.dot(v1)
        dotP0 = intersection_vector.dot(v0)
        dotP1 = intersection_vector.dot(v1)

        inv_den = 1 / (dot00 * dot11 - dot01 * dot01)

        alpha = (dot11 * dotP0 - dot01 * dotP1) * inv_den
        beta = (dot00 * dotP1 - dot01 * dotP0) * inv_den
        gamma = 1 - alpha - beta

        if 0 <= alpha <= 1 and 0 <= beta <= 1 and 0 <= gamma <= 1 and t > 0:
            return True, t
        return False, 0

    def get_normal(self, interception_point: Point | None = None) -> Vector:
        return self.normal
